node/impl/full/eth_trace.go

```diff
@@ -186,53 +192,44 @@ func buildTraces(ctx context.Context, traces *[]*ethtypes.EthTrace, parent *etht
 			}
 		}
 
-		// Handle EVM call special casing
-		//
-		// Any outbound call from an EVM actor on methods 1-1023 are side-effects from EVM instructions
-		// and should be dropped from the trace.
-		if builtinactors.IsEvmActor(parent.Action.FilecoinCodeCid) &&
-			et.Msg.Method > 0 &&
-			et.Msg.Method <= 1023 {
-			log.Debugf("found outbound call from an EVM actor on method 1-1023 method:%d, code:%s, height:%d", et.Msg.Method, parent.Action.FilecoinCodeCid.String(), height)
-			// TODO: if I handle this case and drop this call from the trace then I am not able to detect delegate calls
-		}
-
-		// EVM -> EVM calls
-		//
-		// Check for normal EVM to EVM calls and decode the params and return values
-		if builtinactors.IsEvmActor(parent.Action.FilecoinCodeCid) &&
-			builtinactors.IsEthAccountActor(et.Msg.CodeCid) &&
-			et.Msg.Method == builtin.MethodsEVM.InvokeContract {
-			log.Debugf("found evm to evm call, code:%s, height: %d", et.Msg.CodeCid.String(), height)
-			input, err := handleFilecoinMethodInput(et.Msg.Method, et.Msg.ParamsCodec, et.Msg.Params)
-			if err != nil {
-				return err
-			}
-			trace.Action.Input = input
-			output, err := handleFilecoinMethodOutput(et.MsgRct.ExitCode, et.MsgRct.ReturnCodec, et.MsgRct.Return)
-			if err != nil {
-				return err
-			}
-			trace.Result.Output = output
-		}
-
 		// Handle delegate calls
 		//
 		// 1) Look for trace from an EVM actor to itself on InvokeContractDelegate, method 6.
 		// 2) Check that the previous trace calls another actor on method 3 (GetByteCode) and they are at the same level (same parent)
 		// 3) Treat this as a delegate call to actor A.
-		if trace.Action.From == trace.Action.To &&
-			trace.Action.FilecoinMethod == builtin.MethodsEVM.InvokeContractDelegate &&
-			len(*traces) > 0 {
-			log.Debugf("found delegate call, height: %d", height)
-			prev := (*traces)[len(*traces)-1]
+		if parent.LastByteCode != nil && trace.Action.From == trace.Action.To &&
+			trace.Action.FilecoinMethod == builtin.MethodsEVM.InvokeContractDelegate {
+			log.Debugf("COND7 found delegate call, height: %d", height)
+			prev := parent.LastByteCode
 			if prev.Action.From == trace.Action.From && prev.Action.FilecoinMethod == builtin.MethodsEVM.GetBytecode && prev.Parent == trace.Parent {
 				trace.SetCallType("delegatecall")
 				trace.Action.To = prev.Action.To
 			}
+		} else {
+			// Handle EVM call special casing
+			//
+			// Any outbound call from an EVM actor on methods 1-1023 are side-effects from EVM instructions
+			// and should be dropped from the trace.
+			if builtinactors.IsEvmActor(parent.Action.FilecoinCodeCid) &&
```

Comment:
Both this case and the `InvokeContractDelegate` are EVM special cases. We should put this check on the outside.